osu.Game/Online/Chat/DrawableLinkCompiler.cs

```diff
@@ -1,57 +1,105 @@
 using osu.Game.Overlays;
+using osu.Game.Overlays.OSD;
 using osuTK;
 
 namespace osu.Game.Online.Chat
 {
     /// <summary>
     /// An invisible drawable that brings multiple <see cref="Drawable"/> pieces together to form a consumable clickable link.
     /// </summary>
-    public partial class DrawableLinkCompiler : OsuHoverContainer
+    public partial class DrawableLinkCompiler : OsuHoverContainer, IHasContextMenu
     {
         /// <summary>
         /// Each word part of a chat link (split for word-wrap support).
         /// </summary>
         public readonly List<Drawable> Parts;
 
-        [Resolved(CanBeNull = true)]
-        private OverlayColourProvider overlayColourProvider { get; set; }
+        public readonly LinkDetails Link;
+
+        [Resolved]
+        private OverlayColourProvider? overlayColourProvider { get; set; }
+
+        [Resolved]
+        private OnScreenDisplay? onScreenDisplay { get; set; }
+
+        [Resolved]
+        private GameHost host { get; set; } = null!;
+
+        [Resolved]
+        private ILinkHandler? linkHandler { get; set; }
 
         public override bool ReceivePositionalInputAt(Vector2 screenSpacePos) => Parts.Any(d => d.ReceivePositionalInputAt(screenSpacePos));
 
         protected override HoverSounds CreateHoverSounds(HoverSampleSet sampleSet) => new LinkHoverSounds(sampleSet, Parts);
 
-        public DrawableLinkCompiler(ITextPart part)
-            : this(part.Drawables.OfType<SpriteText>())
+        public DrawableLinkCompiler(ITextPart part, LinkDetails link)
+            : this(part.Drawables.OfType<SpriteText>(), link)
         {
         }
 
-        public DrawableLinkCompiler(IEnumerable<Drawable> parts)
+        public DrawableLinkCompiler(IEnumerable<Drawable> parts, LinkDetails link)
         {
             Parts = parts.ToList();
+            Link = link;
         }
 
         [BackgroundDependencyLoader]
         private void load(OsuColour colours)
         {
             IdleColour = overlayColourProvider?.Light2 ?? colours.Blue;
+
+            Action ??= () =>
+            {
+                Debug.Assert(Link.Action != LinkAction.Custom);
+
+                if (linkHandler != null)
+                    linkHandler.HandleLink(Link);
+                // fallback to handle cases where OsuGame is not available, ie. tournament client.
+                else if (Link.Action == LinkAction.External)
+                    host.OpenUrlExternally(Link.Argument.ToString());
+            };
         }
 
         protected override IEnumerable<Drawable> EffectTargets => Parts;
 
+        public MenuItem[] ContextMenuItems
+        {
+            get
+            {
+                string? url = MessageFormatter.GetUrl(Link);
+
+                if (url == null)
+                    return Array.Empty<MenuItem>();
+
+                return new MenuItem[]
+                {
+                    new OsuMenuItem("Open", MenuItemType.Highlighted, Action),
```

Comment:
I've reverted the logic that decides the action text based on the destination of the link (https://github.com/ppy/osu/pull/23334/commits/e92572b62792b92fef9803a85293000b82f5836a#diff-fe2f5f4b0275915dcaadd5f4eedb0c524d17e906699d64362bdcd828de8c4055L90-L104), because it looked pretty much arbitrary.

Reply:
I'm going to have a rather large issue with all of this I'm afraid. Previously `DrawableLinkCompiler` was a simple wrapper for flowing text sprites to look like a link. Now it's resolving OSDs and handling clipboard and adding link-type-specific logic.

None of this should be here. It should be shipped off to a place specific for "URL-based hyperlinks".